runelite-client/src/main/java/net/runelite/client/plugins/timestamp/TimestampPlugin.java

```diff
@@ -0,0 +1,151 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.timestamp;
+
+import com.google.common.eventbus.Subscribe;
+import java.awt.Color;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoField;
+import java.util.Hashtable;
+import javax.inject.Inject;
+import net.runelite.api.Client;
+import net.runelite.api.Varbits;
+import net.runelite.api.events.ConfigChanged;
+import net.runelite.api.events.ScriptCallbackEvent;
+import net.runelite.api.events.SetMessage;
+import net.runelite.client.callback.ClientThread;
+import net.runelite.client.config.ChatColorConfig;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.client.util.ColorUtil;
+
+@PluginDescriptor(
+	name = "Timestamp",
+	description = "Add timestamps to messages"
+)
+public class TimestampPlugin extends Plugin
+{
+	private static final String WELCOME_MESSAGE = "Welcome to RuneScape.";
+
+	private Color timestampColour;
+	private Hashtable<Integer, String> timestamps = new Hashtable<>();
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private ChatColorConfig chatColorConfig;
+
+	@Inject
+	private ClientThread clientThread;
+
+	@Subscribe
+	public void onSetMessage(SetMessage setMessage)
+	{
+		if (setMessage.equals(WELCOME_MESSAGE))
+		{
+			timestamps.clear();
+		}
+	}
+
+	@Subscribe
+	public void onScriptCallbackEvent(ScriptCallbackEvent event)
+	{
+		int[] intStack = client.getIntStack();
+		int inStackSize = client.getIntStackSize();
+
+		String[] stringStack = client.getStringStack();
+		int stringStackSize = client.getStringStackSize();
+
+		final ZonedDateTime time = ZonedDateTime.now();
+
+		final String dateFormat = time.get(ChronoField.HOUR_OF_DAY) + ":" +
```

Comment:
Move the time/formatting into the getMessageInfo event so it isn't done for each script event.